File: composeApp/src/androidMain/kotlin/com/bimm/takehomeassignmnent/presentation/SakeShopsViewModel.kt
package com.bimm.takehomeassignmnent.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bimm.takehomeassignmnent.domain.SakeShopsRepository
import com.bimm.takehomeassignmnent.state.SakeShopState
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * ViewModel responsible for loading and exposing the list of sake shops.
 *
 * Holds a [SakeShopState] that represents the current UI state:
 * - [SakeShopState.NotInitialized]: nothing has been loaded yet.
 * - [SakeShopState.Loading]: a load is in progress.
 * - [SakeShopState.Success]: data loaded successfully.
 * - [SakeShopState.Error]: an error occurred during load.
 *
 * @property sakeShopsRepository repository used to fetch the sake shops
 */
class SakeShopsViewModel(
    private val sakeShopsRepository: SakeShopsRepository
) : ViewModel() {

    private val _state = MutableStateFlow<SakeShopState>(SakeShopState.NotInitialized)
    /** Publicly exposed UI state flow for observing from the UI. */
    val state: StateFlow<SakeShopState> get() = _state

    /**
     * Initiates loading of sake shops.
     *
     * Emits [SakeShopState.Loading] after a brief delay to simulate network latency,
     * then performs the fetch on [Dispatchers.IO].
     * On success, emits [SakeShopState.Success] with the fetched list.
     * On failure, emits [SakeShopState.Error] with the error message.
     */
    fun loadData() {
        _state.value = SakeShopState.Loading

        viewModelScope.launch {
            delay(1500) // Simulate network delay
            withContext(Dispatchers.IO) {
                try {
                    val shops = sakeShopsRepository.getSakeShops()
                    _state.value = SakeShopState.Success(shops)
                } catch (e: Exception) {
                    _state.value = SakeShopState.Error(e.message ?: "Unknown error")
                }
            }
        }
    }
}

File: iosApp/iosApp/SakeShops/SakeShopsViewModel.swift
//
//  SakeShopsViewModel.swift
//  iosApp
//
//  Created by Marco Antonio Martinez Gutierrez on 25/05/25.
//

import Foundation
@preconcurrency import Shared

/// ViewModel responsible for loading and exposing the list of sake shops.
///
/// Uses an injected `SakeShopsRepository` to fetch data asynchronously
/// and publishes its loading, success, or error state to the UI.
@MainActor
class SakeShopsViewModel: ObservableObject {
    /// Represents the current loading state of the view.
    @Published private(set) var state: SakeShopState = .notInitialized

    /// The repository used to fetch sake shops data.
    private let repository: SakeShopsRepository

    /// Initializes the ViewModel with the provided repository.
    ///
    /// - Parameter repository: The repository responsible for fetching shop data.
    init(repository: SakeShopsRepository) {
        self.repository = repository
    }

    /// Loads the sake shops, simulating a network delay,
    /// and updates the `state` property accordingly.
    ///
    /// Transitions through:
    /// 1. `.loading` when the load begins.
    /// 2. `.success` with the list of shops on successful fetch.
    /// 3. `.error` with an error message if fetching fails.
    func loadData() {
        state = .loading

        Task {
            do {
                // Simulate network latency
                try await Task.sleep(nanoseconds: 1_500_000_000)

                // Fetch shops from repository
                let shops = try await repository.getSakeShops()
                self.state = .success(shops)
            } catch {
                // Publish error state with localized message
                self.state = .error(error.localizedDescription)
            }
        }
    }
}
